sklearn/model_selection/_split.py

```diff
@@ -1865,6 +1866,80 @@ def get_n_splits(self, X=None, y=None, groups=None):
         return len(self.unique_folds)
 
 
+class BinnedStratifiedKFold(StratifiedKFold):
+    """Stratified K-Folds cross-validator
+
+    Provides train/test indices to split data in train/test sets.
+
+    This cross-validation object is a variation of KFold that returns
+    stratified folds by binning a regression target. The folds are made
+    by approximately preserving the distribution of targets in each bin.
+
+    Read more in the :ref:`User Guide <cross_validation>`.
+
+    Parameters
+    ----------
+    n_splits : int, default=5
+        Number of folds. Must be at least 2.
+
+    shuffle : boolean, default=False
+        Whether to shuffle each stratification of the data before splitting
+        into batches.
+
+    n_bins : int, default=10
+        How many quantile bins to use.
+
+    random_state : None, int or RandomState
+        When shuffle=True, pseudo-random number generator state used for
+        shuffling. If None, use default numpy RNG for shuffling.
+
+    Examples
+    --------
+    >>> from sklearn.model_selection import BinnedStratifiedKFold
+    >>> y = np.arange(11.0)
+    >>> np.random.seed(0)
+    >>> np.random.shuffle(y)
+    >>> X = y + 0.1* np.random.randn(len(y))
+    >>> cv = BinnedStratifiedKFold(n_splits=3)
+    >>> skf = cv.split(X, y)
+    >>> print(cv)
+    BinnedStratifiedKFold(n_bins=5, n_splits=3, random_state=None...)
+    >>> indarr = np.zeros(len(y), dtype=bool)
+    >>> for train_index, test_index in skf:
+    ...    print("TRAIN:", train_index, "TEST:", test_index)
+    ...    X_train, X_test = X[train_index], X[test_index]
+    ...    y_train, y_test = y[train_index], y[test_index]
+    TRAIN: [ 5  6  7  8  9 10] TEST: [0 1 2 3 4]
+    TRAIN: [ 0  1  2  3  4  9 10] TEST: [5 6 7 8]
+    TRAIN: [0 1 2 3 4 5 6 7 8] TEST: [ 9 10]
+
+    Notes
+    -----
+    All the folds have size floor(n_samples / n_splits) or
+    floor(n_samples / n_splits) +1,
+    the length is assigned randomly (even if no shuffling is requested)
+    to balance the variance between folds.
+
+    See also
+    --------
+    StratifiedKFold: stratified k-fold generator for classification data
+    """
+
+    def __init__(self, n_splits=5, shuffle=False, n_bins=5,
+                 random_state=None):
+        super().__init__(n_splits, shuffle, random_state)
+        self.n_bins = n_bins
+        if n_bins < 2:
+            raise ValueError("Need at least two bins, got {}.".format(
+                n_bins))
+
+    def _make_test_folds(self, X, y):
+        percentiles = np.percentile(
+            y, np.linspace(0, 100, self.n_bins))
```

Comment:
this should be self.n_bins + 1? (we need n+1 points if we want n bins)

Reply:
I'm not sure. Check out digitize: https://docs.scipy.org/doc/numpy/reference/generated/numpy.digitize.html

It might actually be n_bins - 1? digitize adds an n-th bin. It looks like it creates a new bin for "to the right" but not for "to the left" which is a bit weird?

Reply:
this is our solution from calibration_curve (n_bins + 1 & add a small value to the last threshold),
```
quantiles = np.linspace(0, 1, n_bins + 1)
bins = np.percentile(y_prob, quantiles * 100)
bins[-1] = bins[-1] + 1e-8
binids = np.digitize(y_prob, bins) - 1
```

Reply:
That seems to be the right fix. Also see https://github.com/numpy/numpy/issues/4217